### PR TITLE
[refactor] Replace std::map with switch case statements in lang_util.cpp

### PR DESCRIPTION
**Pull request by aryansoman**
Related issue = #1750

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

Hello, I tried refactoring the first two functions. Once these changes look good, I'll apply them to every function with std::map.

### CODE REVIEW

**Comment by aryansoman** [comment]
I've refactored the functions and put TI_NOT_IMPLEMENTED in the default blocks, and corrected the sizing errors. Let me know if anything else needs work. Thank you!

**Comment by yuanming-hu** [comment]
Very cool! Could you fix the compilation errors? https://github.com/taichi-dev/taichi/runs/1018642631#step:5:200

It's recommended to have a local Taichi built from source: https://taichi.readthedocs.io/en/latest/dev_install.html#developer-installation

File: lang_util_backup.cpp [hunk 1]
```diff
@@ -0,0 +1,519 @@
+// Definitions of utility functions and enums
```
Comment (archibate) [inline lang_util_backup.cpp]: Let's move this file to `misc/lang_util_backup.cpp` to prevent confusion.

**Comment by aryansoman** [comment]
@yuanming-hu The switch-case statement doesn't work for pairs of DataTypes, so should I leave the promoted_type method as-is?

**Comment by yuanming-hu** [comment]
> @yuanming-hu The switch-case statement doesn't work for pairs of DataTypes, so should I leave the promoted_type method as-is?

Sure, sounds good to me! We can leave it to a future PR. Thank you.

File: taichi/lang_util.cpp [hunk 2]
```diff
@@ -106,77 +110,102 @@ std::string data_type_format(DataType dt) {
...
+    case DataType::unknown:
+      return -1;
+
+#define REGISTER_DATA_TYPE(i, j) \
+  case DataType::i:              \
+    return sizeof(j);
+
+      REGISTER_DATA_TYPE(f32, float32);
+      REGISTER_DATA_TYPE(f64, float64);
+      // REGISTER_DATA_TYPE(i8, bool); // duplicate value
```
Comment (yuanming-hu) [inline taichi/lang_util.cpp:127]:
```suggestion
```
Nice catch! Let's remove this line then :-)

File: taichi/lang_util.cpp [hunk 3]
```diff
@@ -196,53 +225,65 @@ std::string binary_op_type_symbol(BinaryOpType type) {
     REGISTER_TYPE(bit_and, &);
     REGISTER_TYPE(bit_or, |);
     REGISTER_TYPE(bit_xor, ^);
-    REGISTER_TYPE(bit_shl, <<);
-    REGISTER_TYPE(bit_sar, >>);
```
Comment (yuanming-hu) [inline taichi/lang_util.cpp:200]: I believe we still need these lines?

**Pull request merged**